nucliadb/src/nucliadb/common/cluster/rebalance.py

```diff
@@ -153,27 +193,189 @@ async def move_set_of_kb_resources(
+):
+    logger.info(
+        "Splitting excess of paragraphs to other shards",
+        extra={
+            "kbid": kbid,
+            "shard": shard.to_dict(),
+        },
+    )
+
+    # First off, calculate if the excess fits in the other shards
+    excess = shard.paragraphs - settings.max_shard_paragraphs
+    other_shards = [s for s in shards if s.id != shard.id]
+    capacity = calculate_capacity(other_shards) * 0.9
+    if excess > capacity:
+        shards_to_add = math.ceil((excess - capacity) / settings.max_shard_paragraphs)
+        logger.info(
+            "More shards needed",
+            extra={
+                "kbid": kbid,
+                "shards_to_add": shards_to_add,
+                "all_shards": [s.to_dict() for s in shards],
+            },
+        )
+        # Add new shards where to rebalance the excess of paragraphs
+        async with (
+            locking.distributed_lock(locking.NEW_SHARD_LOCK.format(kbid=kbid)),
+            datamanagers.with_rw_transaction() as txn,
+        ):
+            sm = get_shard_manager()
+            for _ in range(shards_to_add):
+                await sm.create_shard_by_kbid(txn, kbid)
+            await txn.commit()
+
+        # Recalculate after having created shards, the active shard is a different one
+        shards = await get_rebalance_shards(kbid)
+
+    # Now, move resoures to other shards as long as we are still over the max
+    with_room = [
+        s for s in shards if s.id != shard.id and s.paragraphs < settings.max_shard_paragraphs * 0.9
+    ]
+    shard_paragraphs = shard.paragraphs
+    i = 0
+    while (shard_paragraphs > settings.max_shard_paragraphs) and i < 500:
+        try:
+            # Get the biggest shard with some space left
+            target_shard = with_room[-1]
+        except IndexError:
+            logger.warning("No target shard found for splitting", extra={"kbid": kbid})
             break
 
-        await move_set_of_kb_resources(context, kbid, largest_shard, smallest_shard)
+        await move_set_of_kb_resources(
+            context, kbid, from_shard_id=shard.id, to_shard_id=target_shard.id, count=20
+        )
+
+        # Recalculate shard paragraphs
+        shards = await get_rebalance_shards(kbid)
+        shard_paragraphs = next(s.paragraphs for s in shards if s.id == shard.id)
+        with_room = [
+            s for s in shards if s.id != shard.id and s.paragraphs < settings.max_shard_paragraphs * 0.9
+        ]
+        i += 1
 
-        rebalanced_shards.add(largest_shard)
 
-        shard_paragraphs = await get_shards_paragraphs(kbid)
+async def merge_shard(
+    context: ApplicationContext, kbid: str, shard: RebalanceShard, shards: list[RebalanceShard]
+):
+    logger.info(
+        "Merging shard",
+        extra={
+            "kbid": kbid,
+            "shard": shard.to_dict(),
+        },
+    )
+    empty_shard = False
+
+    i = 0
+    while i < 1000:
```

Comment:
If i=1000 the max number of movements per shard, could you create a constant for better readability?